lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1434 @@
+  bool resetValueUsed = false;
+
+  for (auto &use : target.getUses()) {
+    Operation *useOp = use.getOwner();
+    builder.setInsertionPoint(useOp);
+    TypeSwitch<Operation *>(useOp)
+        // Insert a mux on the value connected to the target:
+        // connect(dst, src) -> connect(dst, mux(reset, resetValue, src))
+        .Case<ConnectOp, PartialConnectOp>([&](auto op) {
+          if (op.dest() != target)
+            return;
+          LLVM_DEBUG(llvm::dbgs() << "  - Insert mux into " << op << "\n");
+          auto muxOp = builder.create<MuxPrimOp>(reset, resetValue, op.src());
+          op.srcMutable().assign(muxOp);
+          resetValueUsed = true;
+        })
+        // Look through subfields.
+        .Case<SubfieldOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubfieldOp>(resetValue, op.fieldIndexAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subindices.
+        .Case<SubindexOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubindexOp>(resetValue, op.indexAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subaccesses.
+        .Case<SubaccessOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubaccessOp>(resetValue, op.index());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        });
+  }
+  return resetValueUsed;
+}
+
+//===----------------------------------------------------------------------===//
+// Reset Network
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// A reset signal.
+///
+/// This essentially combines the exact `FieldRef` of the signal in question
+/// with a type to be used for error reporting and inferring the reset kind.
+struct ResetSignal {
+  ResetSignal(FieldRef field, FIRRTLType type) : field(field), type(type) {}
+  bool operator<(const ResetSignal &other) const { return field < other.field; }
+  bool operator==(const ResetSignal &other) const {
+    return field == other.field;
+  }
+  bool operator!=(const ResetSignal &other) const { return !(*this == other); }
+
+  FieldRef field;
+  FIRRTLType type;
+};
+
+/// A connection made to or from a reset network.
+///
+/// These drives are tracked for each reset network, and are used for error
+/// reporting to the user.
+struct ResetDrive {
+  /// What's being driven.
+  ResetSignal dst;
+  /// What's driving.
+  ResetSignal src;
+  /// The location to use for diagnostics.
+  Location loc;
+};
+
+/// A list of connections to a reset network.
+using ResetDrives = SmallVector<ResetDrive, 1>;
```

Comment:
This is intentional because at the beginning of reset tracing, there are many small reset networks with only a few drives. This tries to cater to that situation to avoid `ResetDrives` occupying a lot of space that is never used in the common case. As the algorithm progresses, only a few reset networks remain, with the number of drives somewhat proportional to the input size, which makes sense to put on the heap anyway.